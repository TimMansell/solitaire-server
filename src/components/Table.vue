<template>
  <table class="table">
    <thead>
      <tr>
        <th
          class="table__cell"
          v-for="(heading, index) in headings"
          :key="index"
        >
          {{ heading }}
        </th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="(row, rowIndex) in items" :key="rowIndex">
        <td
          class="table__cell"
          v-for="(cell, cellIndex) in row"
          :key="cellIndex"
        >
          {{ cell }}
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
export default {
  name: 'Table',
  props: {
    headings: {
      type: Array,
      default: () => [],
    },
    items: {
      type: Array,
      default: () => [],
    },
  },
};
</script>

<style scoped lang="scss">
.table {
  &__cell {
    border: 1px solid var(--bdr-primary);
    padding: var(--pd-sm);
  }
}
</style>
