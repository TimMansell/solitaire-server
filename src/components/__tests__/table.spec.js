import { shallowMount } from '@vue/test-utils';
import Table, { isRowHighlighted } from '@/components/Table.vue';

const propsData = {
  headings: ['Heading 1', 'Heading 2', 'Heading 3'],
  items: [
    [1, 2, 3],
    [4, 5, 6],
  ],
};

describe('Table.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(Table, {
      propsData,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot - placeholder rows', () => {
    const wrapper = shallowMount(Table, {
      propsData: {
        ...propsData,
        items: [],
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should return highlighted row', () => {
    const result = isRowHighlighted([1, 2, 3], { key: '0', value: 1 });

    expect(result).toBe(true);
  });

  it('should not return highlighted row', () => {
    const result = isRowHighlighted([1, 2, 3], { key: '0', value: 2 });

    expect(result).toBe(false);
  });

  it('should show placeholder rows', () => {
    const wrapper = shallowMount(Table, {
      propsData: {
        ...propsData,
        items: [],
      },
    });

    expect(wrapper.find('[data-test="table-placeholder-row"]').exists()).toBe(
      true
    );

    expect(wrapper.find('[data-test="table-row"]').exists()).toBe(false);
  });

  it('should not show placeholder rows', () => {
    const wrapper = shallowMount(Table, {
      propsData,
    });

    expect(wrapper.find('[data-test="table-placeholder-row"]').exists()).toBe(
      false
    );

    expect(wrapper.find('[data-test="table-row"]').exists()).toBe(true);
  });

  it('should show highlight correct row', () => {
    const wrapper = shallowMount(Table, {
      propsData: {
        ...propsData,
        toHighlight: { key: '0', value: 1 },
      },
    });

    expect(
      wrapper.findAll('[data-test="table-row"]').at(0).classes()
    ).toContain('table__row--highlighted');
  });
});
