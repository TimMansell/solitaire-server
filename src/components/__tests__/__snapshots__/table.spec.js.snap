// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.vue matches snapshot - placeholder rows 1`] = `
<table class="table">
  <thead>
    <tr>
      <th class="table__cell">
        Heading 1
      </th>
      <th class="table__cell">
        Heading 2
      </th>
      <th class="table__cell">
        Heading 3
      </th>
    </tr>
  </thead>
  <tbody>
    <tr data-test="table-placeholder-row">
      <td class="table__cell">
        <skeletonloader-stub></skeletonloader-stub>
      </td>
      <td class="table__cell">
        <skeletonloader-stub></skeletonloader-stub>
      </td>
      <td class="table__cell">
        <skeletonloader-stub></skeletonloader-stub>
      </td>
    </tr>
  </tbody>
  <!---->
</table>
`;

exports[`Table.vue matches snapshot 1`] = `
<table class="table">
  <thead>
    <tr>
      <th class="table__cell">
        Heading 1
      </th>
      <th class="table__cell">
        Heading 2
      </th>
      <th class="table__cell">
        Heading 3
      </th>
    </tr>
  </thead>
  <!---->
  <tbody>
    <tr data-test="table-row" class="">
      <td data-test="table-cell" class="table__cell">
        1
      </td>
      <td data-test="table-cell" class="table__cell">
        2
      </td>
      <td data-test="table-cell" class="table__cell">
        3
      </td>
    </tr>
    <tr data-test="table-row" class="">
      <td data-test="table-cell" class="table__cell">
        4
      </td>
      <td data-test="table-cell" class="table__cell">
        5
      </td>
      <td data-test="table-cell" class="table__cell">
        6
      </td>
    </tr>
  </tbody>
</table>
`;
